libvast/src/arrow_table_slice.cpp

```diff
@@ -459,9 +461,20 @@ class row_picker {
     }
   }
 
+  static bool valid(const arrow::StructArray& arr, int64_t row) {
+    auto struct_length = arr.length();
+    for (const auto& field : arr.fields()) {
+      if (field->length() < struct_length || field->length() < row)
+        return false;
+    }
+    return true;
+  }
+
   void operator()(const arrow::StructArray& arr, const legacy_record_type& t) {
     if (arr.IsNull(row_))
       return;
+    if (!valid(arr, row_))
```

Comment:
Shouldn't we print an error here, like `database contains corrupted data for id range [xxxx - yyy)`?

Also, is this only in the export path? (Ie. should we be concerned about performance for this check?)

Reply:
Wouldn't it make more sense to make this an assertion rather than trying to verify data here? I think doing data integrity checks should not be part of exporting.

Reply:
Do we have a mechanism to rate-limit such error logs? I fear this can become spammy on bulk-exports otherwise.

In general this affects every `at()` into such slices, I think the check is in the right place. We should definitely check the performance of this. In principle we could do test for the error condition during table slice materialization and set a flag in the hidden state instead.

Reply:
> Wouldn't it make more sense to make this an assertion rather than trying to verify data here? I think doing data integrity checks should not be part of exporting.

The reality is that databases with incorrect table slices exist. This addition prevents segfaults when they get accessed, an assertion would not help in this case.

Reply:
Don't they only exist for users that used a build that was never released (i.e., from 2-3 weeks ago)?

Reply:
> Don't they only exist for users that used a build that was never released (i.e., from 2-3 weeks ago)?

Only if you assume that nobody created their own schemas with `list<record>` or `map<record>`.

Reply:
For measuring in the hot path on export:

```diff
diff --git a/libvast/src/table_slice.cpp b/libvast/src/table_slice.cpp
index 8cf19bfba..cbac8d110 100644
--- a/libvast/src/table_slice.cpp
+++ b/libvast/src/table_slice.cpp
@@ -706,6 +706,7 @@ filter(const table_slice& slice, expression expr, const ids& hints) {
       result.emplace_back(field.type);
     return result;
   }();
+  auto start = std::chrono::steady_clock::now();
   for (auto id : select(selection)) {
     VAST_ASSERT(id >= offset);
     auto row = id - offset;
@@ -718,6 +719,12 @@ filter(const table_slice& slice, expression expr, const ids& hints) {
       }
     }
   }
+  auto stop = std::chrono::steady_clock::now();
+  auto elapsed
+    = std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(
+      stop - start);
+  VAST_ERROR("filter {}; n = {}, elapsed = {}, elapsed / n = {}", __LINE__,
+             rank(selection), elapsed, elapsed / rank(selection));
   if (builder->rows() == 0)
     return std::nullopt;
   if (builder->rows() == slice.rows())
```

Reply:
This turned out to be a costly check (worst case ~8% slower), so we decided to remove the workaround as discussed above.